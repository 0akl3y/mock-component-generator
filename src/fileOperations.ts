import { Glob } from 'glob'
import { generateMock, MockGeneratorOptions } from './generateMocks'
const glob = require('glob')
const path = require('path')

// eslint-disable-next-line @typescript-eslint/no-var-requires
const fs = require('fs')
// const path = require("path");

const mockFolderExists = (dir: string) => {
  try {
    fs.readdirSync(`${dir}/__mocks__`, {
      encoding: 'utf-8',
    })
    return true
  } catch {
    return false
  }
}

const mockFileExists = (dir: string, filenanme: string) => {
  try {
    const mockFiles: string[] = fs.readdirSync(`${dir}/__mocks__`, {
      encoding: 'utf-8',
    })
    return mockFiles.includes(filenanme)
  } catch {
    return false
  }
}

export async function generateMocksInDir(
  file: string,
  options?: MockGeneratorOptions
) {
  const files = glob.sync(file)

  files.forEach((f) => {
    const fullPath = path.relative(process.cwd(), f)
    const currPath = path.dirname(fullPath)

    const currFile = path.basename(fullPath)
    if (!mockFolderExists(currPath)) {
      fs.mkdirSync(`${currPath}/__mocks__`)
    }
    const content = fs.readFileSync(`${f}`, {
      encoding: 'utf-8',
    })
    const mockContent = generateMock(content, options)

    //It should not overwrite existing mocks
    if (!mockFileExists(file, f) && mockContent) {
      fs.writeFileSync(`${currPath}/__mocks__/${currFile}`, mockContent, {
        encoding: 'utf-8',
      })
    }
  })
}
