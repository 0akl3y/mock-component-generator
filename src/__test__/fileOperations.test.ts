/* eslint-disable @typescript-eslint/no-var-requires */

import { generateMocksInDir } from '../fileOperations'

const fs = require('fs')

const mockFolderPath = `${__dirname}/__mockFilesystem__/__mocks__`

describe('fileOperations', () => {
  beforeAll(() => {
    fs.rmdirSync(mockFolderPath, { force: true, recursive: true })
  })

  it('walks generates mock files for all existing files in the dir', async () => {
    await generateMocksInDir(`${__dirname}/__mockFilesystem__`)

    const mockFiles: string[] = fs.readdirSync(mockFolderPath, {
      encoding: 'utf-8',
    }) as string[]

    mockFiles.forEach((mFile) => {
      const newContent = fs.readFileSync(`${mockFolderPath}/${mFile}`, {
        encoding: 'utf-8',
      })

      expect(newContent).toMatchSnapshot()
    })
  })
})
