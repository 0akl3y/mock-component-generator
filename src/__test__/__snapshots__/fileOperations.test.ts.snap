// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fileOperations walks generates mock files for all existing files in the dir 1`] = `
"import React from 'react'
export const SomeClassComponent = props => {
  return React.createElement(\\"SomeClassComponent\\", props);
};"
`;

exports[`fileOperations walks generates mock files for all existing files in the dir 2`] = `
"import React from 'react'
export const HelloComponent = props => React.createElement(\\"HelloComponent\\", props);"
`;

exports[`fileOperations walks generates mock files for all existing files in the dir 3`] = `
"import React from 'react'
export const foo = jest.fn();"
`;
